Add REST support.

harsh-98 opened the pull request:
Currently for keystore, avm , platform and admin module. Once you guys approach this, I can very easily add support for health , metric and other endpoints.

Solves https://github.com/ava-labs/gecko/issues/211.

I thought of two methods of solving this.
First splitting the uri and creating keystore.CreateUser from /api/keystore/user/create
```
 uriSections := strings.SplitN(strings.ToLower(req.RequestURI), "/", 5)
	// method := fmt.Sprintf("%s.%s%s", uriSections[2], strings.Title(uriSections[4]), strings.Title(uriSections[3]))
```
But the problem with this is some functions have weird name and can't be defined with above approach. So, i added mapping feature. like.

```
var mapping map[string]string = map[string]string{
	"/user/create": "CreateUser",
	"/user/delete": "DeleteUser",
	"/user/list":   "ListUsers",
	"/user/import": "ImportUser",
	"/user/export": "ExportUser",
}
```

It works with alias too as aliases are added via `AddAlias` in router.go. And it uses addforceRouter to add the handler for the aliased endpoint. 
```
Support for keystore, avm, platform and admin is added.
```
For REST endpoints, baseURL is replaced by restBaseURL. So that for platform module, `/ext/P` for rpc and  `/rest/P` is for REST. For endpoint and corresponding aliases, request will be routed to same handler. If you url is `/rest/P/tx/issue`  is  handled similar to `/rest/platform/tx/issue` or `/rest/bc/P/tx/issue` or `/rest/bc/platform/tx/issue`.

harsh-98 commented:
I defined a generic approach for handling the rest urls. the url stating with `/rest/` will be handled by rest_codec. Else the traffic is sent to rpc, like `/ext`.

Example:
```
curl -X POST 127.0.0.1:9650/api/keystore/user/create -d '{"username": "xx", "password":"#Rest12345"}' -H 'content-type:application/json' 
```
Output:
```
{"success":true}
```
Reference:
https://www.gorillatoolkit.org/pkg/rpc
https://github.com/gorilla/rpc/tree/master/v2

collincusce commented:
> I defined a generic approach for handling the rest urls. the url stating with /rest/ will be handled by rest_codec. Else the traffic is sent to rpc, like /ext.

Sooo this part is not going to make us happy and isn't what I was looking for. The "ext" stands for extension and I want that consistent. You can't change our URL format. The specs definitely said we add to the end of the endpoint.

collincusce commented:
Oh just out of curiosity since I dont have time to dig through the code, what is the behavior if the endpoint doesn't have a JRPC? VMs do not have to have JRPCs. The RPC choices are strictly on the VM-level. How does that work?

harsh-98 commented:
I have written a script which does all the steps of https://medium.com/avalabs/how-to-join-avas-denali-test-network-9bbfb353207b, using the REST endpoints defined above. Please give it a try. https://gist.github.com/harsh-98/ceaf6fa92e95269a532916bcdf339c72. The  default network id is used so the fund are there in `ewoqjP7PxY4yr3iLTpLisriqt94hdyDFNgchSxGGztUrTXtNN`

harsh-98 commented:
> Oh just out of curiosity since I dont have time to dig through the code, what is the behavior if the endpoint doesn't have a JRPC? VMs do not have to have JRPCs. The RPC choices are strictly on the VM-level. How does that work?

If module has a handler that is registered with   api.ApiServer it will work. I added the support for avm, platform user and admin. The above script also show case that.

harsh-98 commented:
> > I defined a generic approach for handling the rest urls. the url stating with /rest/ will be handled by rest_codec. Else the traffic is sent to rpc, like /ext.
> 
> Sooo this part is not going to make us happy and isn't what I was looking for. The "ext" stands for extension and I want that consistent. You can't change our URL format. The specs definitely said we add to the end of the endpoint URL.

I haven't touched the old RPC urls , I added that apart form /ext also add `/rest` routes to router and handle the request. The `/ext` routes to rpc. I kept in mind that, changing previous code to large extent mine break things unexpectedly.  You can see only 6 lines are delete , even that because i added the restCodec which handles both /ext and /rest.

harsh-98 commented:
@collincusce I see you are thinking that i only added codec, not the handler for the new urls. It is not the case, I added the new handler at line 100 in the router.go.

collincusce commented:
First off, I love your work, ethusiasm, and speed. This is amazing, and I'm glad you could come up with something so fast!

> I haven't touched the old RPC urls , I added new paths /rest/* which takes structure similar to /ext/* except the prefix /ext. And added those /rest URIs to router and handled the request. The /ext routes to rpc. I kept in mind that, changing previous code to large extent might break things unexpectedly. You can see only 6 lines are deleted , even that because i added the restCodec which handles both /ext and /rest.

We need to keep /ext/ is what I'm saying. The beginning of the URL is not the place to change things for this. That can't be modified. 

Your approach is:

```
127.0.0.1:9650/rest/X/asset/description
```

The correct approach would be:

```
127.0.0.1:9650/ext/X/rest/asset/description
```

> Do you want path as /ext/<module>/rest/<extra path like /user/list>?

Yes. :-) 

I'll deep dive through this as I can. I'm still not certain about the Codec stuff. I'll look at it more closely soon, probably tomorrow. Don't want you to keep wating on me. It's not a race, so if you want to keep adjusting, go for it :100:

harsh-98 commented:
I have  change the rest format to /ext/<`module`>/rest<`extra path`>. I have updated the denali node setup with this format of rest endpoints [script](https://gist.github.com/harsh-98/ceaf6fa92e95269a532916bcdf339c72).

In the last commit, I added these changes:
- Rest URI are of format /ext/<`module`>/rest/<`restURI`>
- If REST request has exact path, then only handle the request. earlier PathPrefix was used.
- Add rest endpoint is optional. And is added via RestEndpoints of common.HTTPHandler

danlaine commented on utils/json/rest_codec.go:
```diff
@@ -0,0 +1,87 @@
+
+//  Checks if the uri is in the mapping of RestCodecRequest
+func (r *RestCodecRequest) Method() (string, error) {
+	//
```
nit: remove `//`

harsh-98 commented:
I have rebased this PR with master branch, done the suggested changes by @danlaine . Also updated the [script](https://gist.github.com/harsh-98/ceaf6fa92e95269a532916bcdf339c72) for demo which setup the node for `denali testnet`.
